lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -2148,14 +2149,51 @@ void TFDeabstraction::formGraphOp(SILTensorOpInfo &opInfo,
         continue;
       }
 
-      // Tensor operand type's are ok.
+      // Handle TensorFlow values. These must be handled separately from
+      // aggregates of TensorFlow values (handled below), so that we do not
+      // generate input lists for them.
       if (isTensorFlowValue(operandTy)) {
         opName +=
           GraphOperationInfo::getInputMarker(GraphOperationInfo::IM_Normal);
         inputs.push_back(operand);
         continue;
       }
 
+      // Remember that we have an input list, this marker is important because
+      // the list may be empty, and we need to know it exists in graph lowering.
+      opName +=
+        GraphOperationInfo::getInputMarker(GraphOperationInfo::IM_InputList);
+      const char *elementMarker =
+        GraphOperationInfo::getInputMarker(GraphOperationInfo::IM_InputListElt);
+
+      // Handle aggregates of TensorFlow values by unpacking them to an input
+      // list.
+      if (isTensorFlowValueOrAggregate(operandTy.getASTType())) {
```

Comment:
`isTensorFlowValueOrAggregate` goes recursively into the aggregate, so it's inefficient to visit the aggregate the second time. How about just recursively try to unpack any aggregate and throwing a diagnostic if there's any non-TF type?